glean/tests/unit/core/metrics/timing_distribution.spec.ts

```diff
@@ -0,0 +1,619 @@
+    const hist = new Histogram(new Functional(2.0, 8.0));
+
+    for (let i = 1; i <= 10; i++) {
+      hist.accumulate(i);
+    }
+
+    const snap = snapshot(hist);
+
+    const expectedJson = {
+      count: 10,
+      sum: 55,
+      values: {
+        "1": 1,
+        "2": 1,
+        "3": 1,
+        "4": 1,
+        "5": 1,
+        "6": 1,
+        "7": 1,
+        "8": 1,
+        "9": 1,
+        "10": 1,
+      },
+    };
+
+    assert.deepEqual(snap, expectedJson);
+  });
+
+  it("can snapshot sparse", function () {
+    const hist = new Histogram(new Functional(2.0, 8.0));
+
+    hist.accumulate(1024);
+    hist.accumulate(1024);
+    hist.accumulate(1116);
+    hist.accumulate(1448);
+
+    const snap = snapshot(hist);
+
+    const expectedJson = {
+      count: 4,
+      sum: 4612,
+      values: {
+        "1024": 2,
+        "1116": 1,
+        "1448": 1,
+      },
+    };
+
+    assert.deepEqual(snap, expectedJson);
+  });
+
+  it("serializer should correctly serialize timing distribution", async function () {
+    const metric = new TimingDistributionMetricType(
+      {
+        category: "aCategory",
+        disabled: false,
+        lifetime: Lifetime.Ping,
+        name: "aTimingDistribution",
+        sendInPings: ["aPing"],
+      },
+      TimeUnit.Nanosecond
+    );
+
+    const id = 100;
+    const startTime = 0;
+    const duration = 60;
+
+    metric.setStart(id, startTime);
+    metric.setStopAndAccumulate(id, duration);
+
+    assert.equal((await metric.testGetValue("aPing"))?.sum, duration);
+  });
+
+  it("set value properly sets the value in all stores", function () {
+    const storesNames = ["aPing", "bPing"];
+
+    const metric = new TimingDistributionMetricType(
+      {
+        category: "aCategory",
+        disabled: false,
+        lifetime: Lifetime.Ping,
+        name: "aTimingDistribution",
+        sendInPings: storesNames,
+      },
+      TimeUnit.Nanosecond
+    );
+
+    const id = 100;
+    const duration = 1;
+
+    metric.setStart(id, 0);
+    metric.setStopAndAccumulate(id, duration);
+
+    storesNames.forEach(async (store) => {
+      const snapshot = await metric.testGetValue(store);
+
+      assert.equal(snapshot?.sum, duration);
+      assert.equal(snapshot?.values[1], 1);
+      assert.equal(snapshot?.count, 1);
+    });
+  });
+
+  it("timing distributions must not accumulate negative values", async function () {
+    const metric = new TimingDistributionMetricType(
+      {
+        category: "aCategory",
+        disabled: false,
+        lifetime: Lifetime.Ping,
+        name: "aTimingDistribution",
+        sendInPings: ["aPing"],
+      },
+      TimeUnit.Nanosecond
+    );
+
+    const id = 100;
+    const duration = 60;
+
+    // Flip around the timestamps, this should result in a negative value which should be
+    // discarded.
+    metric.setStart(id, duration);
+    metric.setStopAndAccumulate(id, 0);
+
+    assert.equal(await metric.testGetValue("aPing"), undefined);
```

Comment:
Above should record an error, right? Let's check for that.

```suggestion
    assert.equal(await metric.testGetValue("aPing"), undefined);
    assert.equal(1, await metric.testGetNumRecordedErrors(ErrorType.InvalidValue));
```

Reply:
Yes, tested this locally and it works as expected.